enterprise/suppress-user/internal/badgerdb/badgerdb.go

```diff
@@ -0,0 +1,279 @@
+package badgerdb
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"sync"
+	"time"
+
+	badger "github.com/dgraph-io/badger/v3"
+	"github.com/dgraph-io/badger/v3/options"
+	"github.com/rudderlabs/rudder-server/enterprise/suppress-user/model"
+	"github.com/rudderlabs/rudder-server/utils/logger"
+	"github.com/samber/lo"
+)
+
+// the key used in badgerdb to store the current token
+const tokenKey = "__token__"
+
+// Opt is a function that configures a badgerdb repository
+type Opt func(*Repository)
+
+// WithSeederSource sets the source of the seed data
+func WithSeederSource(seederSource func() (io.Reader, error)) Opt {
+	return func(r *Repository) {
+		r.seederSource = seederSource
+	}
+}
+
+// WithMaxSeedWait sets the maximum time to wait for the seed to complete.
+// If the seed takes longer than this, the repository will be started in restoring state and all
+// repository methods will return [ErrRestoring] until the seed completes. The default wait time is 10 seconds.
+func WithMaxSeedWait(maxSeedWait time.Duration) Opt {
+	return func(r *Repository) {
+		r.maxSeedWait = maxSeedWait
+	}
+}
+
+// Repository is a repository backed by badgerdb
+type Repository struct {
```

Comment:
This was my initial approach, i.e. to expose a `Start` method on the repository and let the caller decide when to start it.
But the actual usage pattern didn't justify it. Additionally, since a repository could be started or not started, I had to put special guards in the repository's methods to check whether the repository was started or not before proceeding, this introduced extra complexity that I didn't like.
Thus, I ended up deciding that the producer function shall always return either a started repository or an error. Seeding is optional and just a few lines long, I found this to be a good compromise :)